### DISCUSSION TITLE
Why are user customizations supposed to be made under `themes/my-doks-theme/` and not in top-level dirs?

### DISCUSSION THREAD

**Discussion by salim-b** [gethyas/hyas#429] [Ideas]
Hyas (or the respective file structure in [create-hyas](https://github.com/gethyas/create-hyas)) [instructs to put user customizations under `themes/my-XYZ-theme/`](https://github.com/gethyas/create-hyas/tree/main/template-doks/themes/my-doks-theme) instead of just using the Hugo project's top-level dirs. IMHO this unnecessarily complicates the setup (two instead of one place for users to customize files). Any particular reason for this?

Moving all the Hugo component folders from `themes/my-XYZ-theme/` two levels up would be more intuitive I think. It would also allow us to cut down module config by [two entries](https://github.com/gethyas/create-hyas/blob/966c1c3dbca1da1e12e00dd4556478ad0ca17cc4/template-doks/config/_default/module.toml#L75-L81) per Hyas theme.

**Comment by h-enk** [comment]
>  Any particular reason for this?

Follow Hugo's [site skeleton](https://gohugo.io/getting-started/directory-structure/#site-skeleton) and keep [themes override things](https://gohugo.io/getting-started/glossary/#theme) separated.

Note that in line with that, there's an integrations directory for integrations overrides. 

I'm not sure what's more intuitive (keep things separated or all in one place) and what setup minimizes possible conflicts best?

**Reply by salim-b** [reply]
> (...) and keep [themes override things](https://gohugo.io/getting-started/glossary/#theme) separated.

I don't think this makes much sense since `create-hyas`' main use is to scaffold a new Hugo project. For the creator of such project there is really no difference between "themes overrides" and their top-level component folders. Instead, the `themes` dir just complicates things for them.

**Reply by salim-b** [reply]
Also note that Hugo's theme concept and hence [theme components](https://gohugo.io/hugo-modules/theme-components/) are pretty much obsoleted by the Hugo modules system (although they still keep working). The Hugo doc lacks behind on this.

**Reply by h-enk** [reply]
> For the creator of such project there is really no difference between "themes overrides" and their top-level component folders. Instead, the themes dir just complicates things for them.

You're right probably /I guess. I recall a Discussion thread that supports your point.

**Reply by h-enk** [reply]
Ok, I'm with you on this. I do think it's a good idea (making things simpler) to only use the top [directories](https://gohugo.io/getting-started/directory-structure/#directories) — without themes 😏

**Reply by salim-b** [reply]
Glad to hear 😄. Do you want to tackle this yourself or would you welcome a PR?

**Reply by pensivedog** [reply]
For folks coming from other systems, such as Drupal, it's definitely more intuitive to have a `themes` directory separated from other site configurations and content. Also, Hugo is set up to allow themes to be broken into [components](https://gohugo.io/hugo-modules/theme-components/) using the `themes` directory, allowing for theme inheritance, which is a familiar practice for people coming from other CMS's that use base themes and subthemes. Here is some [additional discussion](https://github.com/gohugoio/hugo/issues/4460) on why theme components was introduced in Hugo.

**Reply by h-enk** [reply]
> For folks coming from other systems, such as Drupal, it's definitely more intuitive to have a themes directory separated from other site configurations and content.

That's exactly why I set it up like so.

> Also, Hugo is set up to allow themes to be broken into [components](https://gohugo.io/hugo-modules/theme-components/) using the themes directory, allowing for theme inheritance,

Yes, the [theme skeleton](https://gohugo.io/getting-started/directory-structure/#theme-skeleton) installs in the `themes` directory — part of the [site skeleton](https://gohugo.io/getting-started/directory-structure/#site-skeleton)

On second/third thought, I think it's best to keep the original setup id est  the `themes` (and `integrations`) folders. You can always use the top level folders if you want that.

**Reply by h-enk** [reply]
See also [Creating new site and theme skeletons with v0.118.0 and later](https://discourse.gohugo.io/t/creating-new-site-and-theme-skeletons-with-v0-118-0-and-later/46018)

**Reply by salim-b** [reply]
> For folks coming from other systems, such as Drupal, it's definitely more intuitive to have a `themes` directory separated from other site configurations and content.

That's why there *are* separate component folders for `config`, `content`, `layouts` etc. in a Hugo project. Hyas just replicates that same folder structure under the `themes` subdir, so users are confronted with two places to put archetypes (either `archetypes/` or `themes/my-doks-theme/archetypes/`), two places to put assets (either `assets/` or `themes/my-doks-theme/assets/`), two places to put layouts (either `layouts/` or `themes/my-doks-theme/layouts/`) etc.

> Also, Hugo is set up to allow themes to be broken into [components](https://gohugo.io/hugo-modules/theme-components/) using the `themes` directory, allowing for theme inheritance, which is a familiar practice for people coming from other CMS's that use base themes and subthemes.

The `themes/` subdir was designed before Hugo's module system was introduced and is pretty much obsoleted by the [`mounts` configuration](https://gohugo.io/hugo-modules/configuration/). Hugo modules allow for everything that the old themes concept did, but in a more granular way.

> On second/third thought, I think it's best to keep the original setup id est the `themes` (and `integrations`) folders. You can always use the top level folders if you want that.

I respectfully disagree. It unnecessarily complicates Hyas' setup and maintenance. I think it's a worthwhile goal to keep it as simple and standard-conformant as possible. But I don't think trying to resemble a certain folder structure from a certain CMS (Drupal) is a promising approach to that.

**Reply by h-enk** [reply]
⚠️  Hyas is a (strongly opinionated) Node.js web framework  (not a Hugo framework) that uses Hugo as a site generator (for its fast website build times and flexible project organization). See also [Why Hyas?](https://docs.gethyas.com/concepts/why-hyas/)

Also, note that following [Joe's instructions to create a functional site and theme](https://discourse.gohugo.io/t/creating-new-site-and-theme-skeletons-with-v0-118-0-and-later/46018)

```bash
hugo new site mysite
cd mysite
hugo new theme mytheme
echo "theme = 'mytheme'" >> hugo.toml
hugo server
```

You'll get the following directory structure

```bash
D:.
├───archetypes
├───assets
├───content
├───data
├───i18n
├───layouts
├───static
└───themes
    └───mytheme
        ├───archetypes
        ├───assets
        │   ├───css
        │   └───js
        ├───content
        │   └───posts
        │       └───post-3
        ├───data
        ├───i18n
        ├───layouts
        │   ├───partials
        │   │   └───head
        │   └───_default
        └───static
```

**Reply by salim-b** [reply]
>  Hyas is a (strongly opinionated) Node.js web framework (not a Hugo framework)  (...)

Sure, I got that.

> Also, note that following [Joe's instructions to create a functional site and theme](https://discourse.gohugo.io/t/creating-new-site-and-theme-skeletons-with-v0-118-0-and-later/46018)

I don't see how Joe's post is an argument for Hyas' current folder structure duplication. He explains how to scaffold a local folder with Hugo's classic theme structure, mainly to build minimal reproducible examples.

What are the advantages for Hyas ***users*** to have the `themes/mytheme/` subfolder besides the project's' top-level folders? I think there are actually *none*, which is my whole point.

**Reply by h-enk** [reply]
>  [..] I think it's a worthwhile goal to keep it as simple and standard-conformant as possible. [..]

Agreed.

To conclude things and make a decision going forward, I think the common current practice should guide us here.

When you come from Drupal or WordPress, the themes and plugins (or integrations) directories make sense.  However, when you come from Hugo or other Node.js frameworks (like e.g. 11ty and Astro) not so.

Since Hyas' main user base will be rooted in Hugo, I think we should follow Hugo's way, id est use the root component folders — only.

> Do you want to tackle this yourself or would you welcome a PR?

I will do this then.

**Reply by pensivedog** [reply]
I'm an amateur and perfectly happy to defer to others in this discussion. My interest is in ensuring the sites I build with Hyas can be easily updated to benefit from future Hyas improvements, and that I can refer to Hugo documentation when working on my Hyas-based sites.

> I think we should follow Hugo's way, id est use the root component folders — only

At a glance, including a `themes` directory seems to be the Hugo way. Other than this thread, is there any official discussion of deprecating the `themes` directory in Hugo?

Hugo's [directory structure](https://gohugo.io/getting-started/directory-structure/) and site skeleton include a `themes` directory and themes are still included in official Hugo documentation. In my understanding, which is admittedly quite limited, the `themes` directory provides modularity by allowing theme inheritance and keeping theme components separate from the rest of a Hugo site. Wouldn't removing the `themes` directory remove some modularity, or make it more complicated to achieve?

In my experience, the `themes` directory also seems to be widely used in community and premium themes.

There's a lot I don't understand about Hugo and best practices of Hugo developers, so my apologies if I've added extraneous detail to this discussion.

**Reply by h-enk** [reply]
The actual Doks theme (Doks core) lives in the  `node_modules/@hyas/doks-core` folder. You could look at it as a Hugo `themes` subfolder (theme) and think of it as a parent theme. The module can easily be updated for future improvements.

The `themes` (and `integrations`) directories were added for customizations/overrides (think child theme) *only* — so not fully in line with the Hugo way.

With the new setup, modularity is still there — at the main level — and more in line with the Hugo way.

**Reply by salim-b** [reply]
> My interest is in ensuring the sites I build with Hyas can be easily updated to benefit from future Hyas improvements,

Your site will be as easily updateable if you use the top-level folders for customizations as it was when using the `themes/` subfolder. In both cases you will automatically benefit from updates to [doks-core](https://github.com/gethyas/doks-core) when running `npm update`, and in both cases you'll need to manually incorporate improvements to the default configuration in [create-hyas](https://github.com/gethyas/create-hyas/tree/main/template-doks), like https://github.com/gethyas/create-hyas/pull/27 for example.

> and that I can refer to Hugo documentation when working on my Hyas-based sites.

Official Hugo documentation should totally apply if you're using top-level dirs.

> At a glance, including a `themes` directory seems to be the Hugo way.

As I explained, not necessarily anymore. It depends on how your site is configured to use the theme. Using the `themes` dir *the (old) Hugo way* would necessarily include setting `theme = [ "my-doks-theme" ]` in your project's main `hugo.toml` config file **and** incorporating the *whole* theme directly in `themes/my-doks-theme/` either via Git submodules or copying files.

Doks opts for a more powerful and user-friendly approach by using [`mounts`](https://gohugo.io/hugo-modules/configuration/) from the more modern Hugo modules infrastructure instead of the themes subfolder and does not use Hugo's `theme` config option.

So no, Doks current folder duplication is not *the Hugo way* as I see it.

> In my experience, the `themes` directory also seems to be widely used in community and premium themes.

Yes, that's true and probably in partial due to how outdated/incomplete Hugo's official docs are. For example the [Theme components](https://gohugo.io/hugo-modules/theme-components/) chapter has an info notice at the top saying

> *This section contain information that may be outdated and is in the process of being rewritten.*

That notice [was added 4 years ago](https://github.com/gohugoio/hugoDocs/blame/master/content/en/hugo-modules/theme-components.md) by the great guy who mainly developed Hugo modules but at the same time didn't care too much about doing the necessary documentation refactoring and resolvement of conceptual conflicts (bep, Hugo's current lead dev). This may also be because Hugo's documentation is to a much further degree a community project with many more contributors than the underlying Hugo Go codebase. And bep seems to be too reticent regarding refactorings of documentation sections he didn't write himself.

Note that the old themes-way of configuring a Hugo site without Hugo modules will most likely continue to work for another decade or more since Hugo devs really like to stay backwards compatible if possible.

> There's a lot I don't understand about Hugo and best practices of Hugo developers, so my apologies if I've added extraneous detail to this discussion.

Don't worry, sober deliberation is always a promising path to insight. 🙂

**Reply by pensivedog** [reply]
Thank you both for clarifying all these details!
